### PR TITLE
Re-organization of backend

### LINKED ISSUE #1521: Add attachments to form
As an eApp applicant I want to be able to attach documents to my SF-86 so that I can include additional items required by my organization when I submit my SF-86

Acceptance Criteria
- [x] User is able to attach documents
- [x] Only PDF type documents can be attached
- [x] User is able to view all attachments attached to the form
- [x] Users can open attachments attached to the form
- [x] User can remove attachments
- [x] Support file size limit of (TBD)
- [x] Attachments must be encrypted in-transit to storage and while persisted
- [x] Stored as blob in the database
- [x] Utilize the layout/behavior from #1572

### PR DESCRIPTION
**Pull request by bmallred**
Work on backend restructuring to make it easier to swap out implementations (such as geocoding, logging, or database services). It also helps to ensure circular dependencies are avoided.

 1. The root package should not depend on any other package
 2. Subpackages should be grouped by dependencies existing between the domain and the implementation
 3. A mock subpackage used for testing
 4. Main packages should bring dependencies together

This also resolves #1521 including the ability to perform CRUD operations on attachments.

### CODE REVIEW

**Comment by bmallred** [comment]
@floresj running through some integration tests and then will be ready!

**Comment by floresj** [comment]
Everything is looking good. I went through and populated all sections with some basic info. Just ran into a few issues with persisting some of the subsections. 

- Your History
  - Where you lived
  - Employment Activities
- Relationships
  - People who know you well
  - Relatives
-  Citizenship
  - Foreign Passports
-  Military History
  - Foreign military

I'm going to run through these again just to make sure I don't have any weird environmental issue.

File: api/collection.go [hunk 1]
```diff
@@ -195,7 +195,8 @@ func (entity *Collection) SetID(id int) {
 
 // collectionItemIDs the Collection item identifiers.
 func (entity *Collection) collectionItemIDs(context DatabaseService) {
-	count := context.Count(&CollectionItem{}, "id = ?", entity.ID)
+	var count int
+	context.CountExpr(&CollectionItem{}, "max(index) as max", &count, "id = ?", entity.ID)
```
Comment (bmallred) [inline api/collection.go]: Forgot the unique index for `collection_items` was **id**, **index**, and **name** and that each item has the same **index** so needed to retrieve the max and go from there.

File: api/cmd/server/main.go [hunk 2]
```diff
@@ -54,6 +54,8 @@ func main() {
 
 	// Declare a new router with any middleware injected
 	r := mux.NewRouter()
+	r.Use(http.LoggingHandler{Log: logger}.Middleware)
```
Comment (bmallred) [inline api/cmd/server/main.go]: `gorilla/mux` now supports middleware but had to update from `v1.1` to `v1.6.1`. The next release will allow passing multiple middleware in one call but this works for right now. If there are multiple middleware functions then they are called in the order they are assigned.

**Comment by floresj** [comment]
Just FYI for others who read this. Be sure to update the glide dependencies to get the latest gorilla mux router.

**Comment by floresj** [comment]
Minimum data being saved in all sections

image

**Comment by floresj** [comment]
Very nice

image

Showing link in menu
image

**Comment by floresj** [comment]
Also validated that it accepts required file types.

**Comment by bmallred** [comment]
Did you try maximum size too? May have to adjust the NGINX config and restart that container.

**Comment by floresj** [comment]
Yessir. Speaking of maximum... this page had so many changes that my chrome kept crashing on my laptop. Couldn't handle all the awesome  changes.

**Pull request merged**